pkg/ruler/evaluator_jitter.go

```diff
@@ -25,13 +33,28 @@ func NewEvaluatorWithJitter(inner Evaluator, maxJitter time.Duration, rngSource
 	return &EvaluatorWithJitter{
 		inner:     inner,
 		maxJitter: maxJitter,
-		rng:       rand.New(rngSource),
+		hasher:    hasher,
 	}
 }
 
 func (e *EvaluatorWithJitter) Eval(ctx context.Context, qs string, now time.Time) (*logqlmodel.Result, error) {
-	jitter := time.Duration(e.rng.Int63n(e.maxJitter.Nanoseconds()))
-	time.Sleep(jitter)
+	time.Sleep(e.calculateJitter(qs))
 
 	return e.inner.Eval(ctx, qs, now)
 }
+
+func (e *EvaluatorWithJitter) calculateJitter(qs string) time.Duration {
+	var h uint32
+
+	// rules can be evaluated concurrently, so we protect the hasher with a mutex
+	e.Lock()
+	{
+		_, _ = e.hasher.Write([]byte(qs))
+		h = e.hasher.Sum32()
+		e.hasher.Reset()
+	}
+	e.Unlock()
+
+	ratio := float32(h) / math.MaxUint32
+	return time.Duration(ratio * float32(e.maxJitter.Nanoseconds()))
```

Comment:
q: How do you make sure returned duration is within `e.maxJitter` here? I'm failing to understand. Am I missing something here?

Reply:
The hasher (which implements [`hash.Hash32`](https://pkg.go.dev/hash#Hash32)) returns a `uint32`, so by dividing it by `math.MaxUint32` we get a ratio from 0 to 1. The ratio can not be >1 so the jitter will always be between `0` and `maxJitter`.